### PR TITLE
[WIP] readability changes to 05Mocks 

### PR DESCRIPTION
**Pull request by mxochicale**
## Addressing #220

### CODE REVIEW

**Comment by alessandrofelder** [comment]
All looks great to me. I am unsure about the best way to merge `.ipynb` files as the diff contains metadata around the cell status (I used `nbdime` for the first time today, to look at the notebook diff). Any suggestions @dpshelio ?

**Review by alessandrofelder — COMMENTED** [review]
Looks good!

**Comment by dpshelio** [comment]
Added nbreview to this repository to allow reviews of notebboks.

File: ch03tests/05Mocks.ipynb [hunk 1]
```diff
@@ -4,54 +4,75 @@
    "cell_type": "markdown",
```
Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>I like to have the definition, as our course teaches to no-native speakers. That helps to make the connections of what we are trying to do.</p><p> </p><p>Real world examples could be added towards the bottom of the notebook, as to guide the students to look more details about it. <a href="https://webaccess.berkeley.edu/ask-pecan/click-here" target="_blank">Meaningful  hyperlinks helps accessibility</a>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>Please, don't number sections. We can get that done if wanted in the post-processing and it will require lest update.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>What does this try to show?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>2.1 section is too large. Break it into a normal text paragraph.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>Using sections are a bit misleading. A paragraph of text, explaining what's next, is preferable.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>Please, try to follow pep8 syntax (<code>call_number = 4</code>) Also, why are we using <code>name</code> if it's empty? use a <code>_</code> instead if we are not using it or explain it.    </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>This text should be placed before the above call, so that we are telling the reader what's comign.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>This cell could be removed because it was exemplified with the last call already.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>Shouldn't this <code>mock_get.mock_calls</code> be within the context manager? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (dpshelio) [inline ch03tests/05Mocks.ipynb:4]: <p>What's the reason to make <code>params</code> an argument for the function? I see you've added an example with a different set of parameters below. But tin the example we are building a test and the parameters need to be the good ones, the ones we want to check the call is being done so when the function is called that way. In other words, the defaults to the <code>map_at</code> function. Showing two set of <code>params</code> in this example diverts the understanding of what's being tested.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (mxochicale) [reply ch03tests/05Mocks.ipynb:4]: "**`unittest.mock` offers a base class for mocking objects called Mock. The following is an example how to define and object, call it and return id.**"

Comment (mxochicale) [reply ch03tests/05Mocks.ipynb:4]: Thanks @dpshelio, 
I think the example in the notebook is testing the use of the right parameters to do a right remote call: " Instead, we can use mocks to assert that our code does the right thing in terms of the messages it sends: the parameters of the function calls it makes to the remote resource". Just amended the cell to show this:

```
def test_build_default_params(_base, _params):
    with patch.object(requests,'get') as mock_get:
        default_map = map_at(51.0, 0.0)
        mock_get.assert_called_with(_base, params=_params)

base="https://static-maps.yandex.ru/1.x/?"
good_params={
    'z':12,
    'size':'400,400',
    'll':'0.0,51.0',
    'lang':'en_US',
    'l': 'map'
}        
test_build_default_params(base, good_params)
```
and

```
base="https://static-maps.yandex.ru/1.x/?"
bad_params={
    'z':15,
    'size':'400,400',    
    'll':'0.0,51.0',
    'lang':'en_US',
    'l': 'map'
}        
test_build_default_params(base, bad_params)
```
In any case, as I am new to this kind of testing. Examples require to be more intuitive and perhaps add more details for patching as there are few handy tests (e.g. https://kimsereylam.com/python/2021/03/19/how-to-use-patch-in-python-unittest.html). Happy to chat about this bits with @alessandrofelder and yourself next Tuesday :) 

Thanks.

File: ch03tests/05Mocks.ipynb [hunk 2]
```diff
@@ -12,51 +12,42 @@
...
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## 1. The Mock Object\n",
-    "**1.1 `unittest.mock` offers a base class for mocking objects called Mock.**"
+    "## The Mock Object\n",
+    "`unittest.mock` offers a base class for mocking objects called Mock. The following is an example how to define and object, call it and return id.\n"
```
Comment (alessandrofelder) [inline ch03tests/05Mocks.ipynb:39]:
```suggestion
    "`unittest.mock` offers a base class for mocking objects called Mock. The following is an example how to define an object, call it and return it.\n"
```
small typos

File: ch03tests/05Mocks.ipynb [hunk 3]
```diff
@@ -359,14 +210,14 @@
...
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "### 3.1 Using mocks to model test resources\n",
+    "### Using mocks to model test resources\n",
     "Often we want to write tests for code which interacts with remote resources. (E.g. databases, the internet, or data files.)\n",
     "We don't want to have our tests *actually* interact with the remote resource, as this would mean our tests failed\n",
```
Comment (alessandrofelder) [inline ch03tests/05Mocks.ipynb:222]:
```suggestion
    "We don't want to have our tests *actually* interact with the remote resource, as this would mean our tests could fail unpredictably from time to time, through no fault of our own, \n",
```
mini-improvement, while we're here :)

File: ch03tests/05Mocks.ipynb [hunk 4]
```diff
@@ -464,93 +294,64 @@
...
+    "test_build_default_params(base, good_params)"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
     "That was quiet, so it passed. \n",
-    "When writing tests, one usually modifies one of the expectations, to something 'wrong', just to check it's not\n",
+    "When writing tests, we usually modifies one of the expectations, to something 'wrong', just to check it's not\n",
```
Comment (alessandrofelder) [inline ch03tests/05Mocks.ipynb:322]:
```suggestion
    "When writing tests, we usually modify one of the expectations, to something 'wrong', just to check it's not\n",
```
small typo

File: ch03tests/05Mocks.ipynb (hunk 1 above)
Comment (alessandrofelder) [reply ch03tests/05Mocks.ipynb:4]: I think a test called `test_build_default_params` should only test the default parameters. But I like the idea of showing the versatility of mocking. Could we have an additional bit of text after `test_build_default_params` that says something like "Mocking can also be used to test how we handle errors we receive from the remote resource" and then have a negative test that checks we handle a mocked dodgy internet connection as expected?

**Review by alessandrofelder — APPROVED** [review]
Some tiny additional comments :)
